Documentation/planning/arcade-powered-source-build/in-arcade.md

```diff
@@ -0,0 +1,121 @@
+# Source-build as a part of the Arcade build
+
+Source-build does a lot of stuff. A sampler:
+
+1. Clones the upstream repos to form a layout.
+1. Sets up package cache and source-built intermediate output directories.
+1. Orchestrates build order.
+1. Provides a number of source-build-specific args to build commands.
+1. Runs extra targets before and after each build to fill in missing
+   source-build compatibility, if necessary.
+1. Builds tarballs--`.tar.gz` file containing everything necessary to build a
+   specific version of the SDK from source without prebuilts.
+1. Maintains scripts specific to the tarballs to enable scenario like
+   bootstrapping.
+1. Track prebuilt usage to ensure repos don't pull in dependencies not built
+   from source.
+1. Automatically decompile and recompile certain dependencies used as prebuilts
+   so that they can be maintained as source code even if we don't have the
+   original source code.
+
+The main doc discusses why this needs to be done in Arcade. We should also aim
+at all of this being doable via a single `./build.sh [...]` command. This allows
+source-build to be easily reproducible with a local build, something the project
+has been sorely missing and bars all but the most committed and awesome
+contributors.
+
+## Additions to the `build.sh` procedure
+
+It seems that source-build can fit into the existing Arcade extension points.
+
+### Tool Restore
+
+This point is normally used to restore tooling that may impact the ordinary
+restore by adding MSBuild targets. Source-build can add a step to get
+source-build dependencies, with several options:
+
+1. Use transport package. Restores the [intermediate nupkg] and sets it up as a
+   restore source.
+2. Bring-your-own dependencies. A set of args allow the user (or automation) to
+   point to existing intermediates on disk.
+3. Build dependencies from source. Instead of downloading intermediates, clone
+   the repos and build them all.
+
+Any of those options allows the source-build of the current repo to continue
+afterwards with source-built intermediates. Then, source-build sets up
+RestoreSources, a Package Version Props, a blob root, and any repo-specific
+special requirements.
+
+At this point in the build, we also have all we need to create a
+source-buildable tarball. We should allow one to be created as early as possible
```

Comment:
Tarball generation would still require an initial production build through all repos, though, for 2 reasons:
1) the reference-only package ilasm/ildasm process
2) Source generation in coreclr.

Also, I'm not sure where the prebuilt detection logic goes in this scenario.

Reply:
I think we should fix these so we don't need a full production build. Reference-only prebuilts should be tracked (so we don't need to discover them via performing a build), source generation in coreclr is trickier but I think we should track it in source-build to formulate ideas. I'll note these in the issue tracking fullbuildless tarball.

> Also, I'm not sure where the prebuilt detection logic goes in this scenario.

Yeah, I neglected this... in terms of ordering it needs to go in Publish to be as sure as possible that all restores for the repo are complete. (Since I don't think all restoring is guaranteed to happen during arcade's Restore phase, unfortunately.)